dynamodb_client/base_table_client.go

```diff
@@ -0,0 +1,529 @@
+	client      dynamodbiface.DynamoDBAPI
+	pk          *PrimaryKey
+	tableName   string
+	serviceName string
+}
+
+// TableClient defines a context aware table based interface to manage a dynamodb
+// table.
+type TableClient interface {
+	// Reads
+	GetItem(ctx context.Context, hashKey, rangeKey string) (DynamoItem, error)
+	// Writes
+	PutItem(ctx context.Context, hashKey, rangeKey string, item DynamoItem) error
+	DeleteItem(ctx context.Context, hashKey, rangeKey string) error
+	// Table management
+	Create(ctx context.Context) error
+	Delete(ctx context.Context) error
+
+	// LOOKING FOR SOMETHING???
+	// This implementation was pulled from another project. The functions above have had tests added.
+	// If you want to use any of the functionality below, please uncomment the declaration and write a test
+	// for it. The implementation is already in this file, though without tests we don't know if it is correct.
+
+	// Reads
+	// Query(ctx context.Context, opts QueryOpts) ([]map[string]*dynamodb.AttributeValue, error)
+	//	BatchGetDocument(ctx context.Context, keys []DynamoKey, consistentRead bool, v interface{}) ([]error, error)
+	// Writes
+	//	ConditionExpressionPutItem(ctx context.Context, hashKey, rangeKey string, item DynamoItem, expression AwsExpression) error
+	//	ConditionExpressionDeleteItem(ctx context.Context, hashKey, rangeKey string, expr AwsExpression) error
+	//	ConditionExpressionUpdateAttributes(ctx context.Context, hashKey, rangeKey string, expr AwsExpression) error
+	//	BatchPutDocument(ctx context.Context, keys []DynamoKey, items []DynamoItem) ([]error, error)
+	//	BatchDeleteDocument(ctx context.Context, keys []DynamoKey) ([]error, error)
+	// Table Management
+	//	UpdateStreamConfiguration(ctx context.Context) error
+}
+
+type TableClientOpts struct {
+	Desc        *dynamodb.TableDescription
+	Client      dynamodbiface.DynamoDBAPI
+	ServiceName string
+}
+
+func NewBaseTableClient(opts TableClientOpts) (*BaseTableClient, error) {
+	tableName := aws.StringValue(opts.Desc.TableName)
+	pk, err := buildPrimaryKey(opts.Desc)
+	if err != nil {
+		return nil, fmt.Errorf("Error building primary key for %s: %v", tableName, err)
+	}
+
+	table := &BaseTableClient{
+		desc:        opts.Desc,
+		client:      opts.Client,
+		pk:          &pk,
+		tableName:   tableName,
+		serviceName: opts.ServiceName,
+	}
+
+	return table, nil
+}
+
+func (dt *BaseTableClient) Query(ctx context.Context, opts QueryOpts) ([]map[string]*dynamodb.AttributeValue, error) {
+	input := dt.buildQuery(opts)
+
+	req, output := dt.client.QueryRequest(input)
+	err := dt.sendWithTracing(ctx, req)
+
+	return output.Items, err
+}
+
+func (dt *BaseTableClient) PutItem(ctx context.Context, hashKey, rangeKey string, item DynamoItem) error {
+	item = dt.addPrimaryKey(hashKey, rangeKey, item)
+
+	input := &dynamodb.PutItemInput{
+		Item:      item,
+		TableName: dt.desc.TableName,
+	}
+
+	req, _ := dt.client.PutItemRequest(input)
+	err := dt.sendWithTracing(ctx, req)
```

Comment:
If https://github.com/remind101/pkg/pull/109 is merged, then we can just remove all of this and use `PutItemWithContext`. To add tracing to the client, with the table name included in the spans:

```golang
s := contrib.WithTracing(session.New())
c := dynamodb.New(s)
c.Handlers.Send.PushBack(func(r *request.Request) {
        span := opentracing.SpanFromContext(r.Context())
        span.SetTag(dd_opentracing.SpanType, "db")
        span.SetTag("aws.table_name", dt.tableName)
})
```